### PR TITLE
[RW-4645][risk=no] CircleCI config upgrade to version 2.1

### PR DESCRIPTION
**Pull request by aweng98**
Upgrade CircleCI yaml to version 2.1, leveraging new v2.1 features like `commands`, `parameters`, `when`. Re-organized common actions into `commands` for reuse by `jobs`.

High-level structures:
Screen Shot 2020-04-25 at 4 54 29 PM

Notes:
-  Kept all ` jobs` names unchanged to avoid adding new confusion. and number of `jobs` are not changed.
-  Use pipeline `parameters` for set/access common configs. 
-  `command` `git_checkout` combines `- checkout` and `- git submodules update`.
-  new cache `command` s. `ui_dependencies`, `e2e_dependencies` and `api_cache` with parameters.
- New anchor `filter_pr`, for running workflow jobs on non-master and non-release (tags) branches. In other words, run workflow's jobs for PR branches after pushed to origin.

I am not able test out changes in `deploy-staging`, `deploy-perf`, `api-deploy-to-test` and `ui-deploy-to-test` because these jobs are filtered by `filter_master`.

### CODE REVIEW

**Comment by jaycarlton** [comment]
@aweng98 Please provide a PR description telling us what you've changed and why, and how you've tested it.

File: .circleci/config.yml [hunk 1]
```diff
@@ -22,76 +27,201 @@ anchors:
...
+
+  ensure_branch_has_api_changes:
+    description: "Ensure branch has changes to API Code"
+    steps:
+      - run:
+          name: Ensure Branch has Changes to API Code
+          command: |
+            if [ ${CIRCLE_BRANCH} != "" ] && [ ${CIRCLE_BRANCH} != "master" ] && [ $(git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | grep api/ | wc -l | xargs) == 0 ]; then
+              echo No relevant changes on api directory in non-master branch. Skipping
+              circleci step halt
```
Comment (jaycarlton) [inline .circleci/config.yml:116]: BTW  I looked for a way to mark the tests as skipped or cancelled instead of passed in this situation. Filed an enhancement request with circle.

File: .circleci/config.yml [hunk 2]
```diff
@@ -101,146 +231,154 @@ jobs:
...
+      - persist_to_workspace:
+          root: .
+          paths:
+            - api
+
+  api-deploy-to-test:
+    executor: workbench_docker
+    steps:
+      # - git_checkout
+      # Use "api-local-test" job's workspace
```
Comment (jaycarlton) [inline .circleci/config.yml:306]: Interesting. I see how this would be faster, but I thought jobs weren't supposed to be order-dependent (though the steps in them are).

File: .circleci/config.yml [hunk 3]
```diff
@@ -264,136 +402,98 @@ jobs:
...
+          project_name: "all-of-us-workbench-test"
+          deploy_arg: "deploy-ui"
+          working_dir: ~/workbench/ui
+
 
   # In staging and perf, we deploy all at once in order to track and control
   # the release for a tagged version only. In "test", we maintain individual
   # release jobs for API and UI to allow faster deploys and because we lack
   # a version tag.
   deploy-staging:
```
Comment (jaycarlton) [inline .circleci/config.yml:425]: Can we make the `deploy_` jobs parameterized by the destination environment (maybe later)?

File: .circleci/config.yml [hunk 4]
```diff
@@ -264,136 +402,98 @@ jobs:
...
-            ../ci/activate_creds.sh circle-sa-key.json
-            ./project.rb deploy \
-              --project all-of-us-rw-perf \
-              --account circle-deploy-account@all-of-us-workbench-test.iam.gserviceaccount.com \
-              --git-version "${CIRCLE_TAG}" \
-              --app-version "${CIRCLE_TAG}" \
-              --circle-url "https://circleci.com/workflow-run/${CIRCLE_WORKFLOW_ID}" \
-              --key-file circle-sa-key.json \
-              --promote
+            sleep 2m
```
Comment (jaycarlton) [inline .circleci/config.yml:468]: I really get nervous seeing hard-coded sleeps. Can we not teach Circle to wait for some event or trigger to proceed?

File: .circleci/config.yml [hunk 5]
```diff
@@ -264,136 +402,98 @@ jobs:
...
+            dockerize -wait tcp://127.0.0.1:4200 -timeout 3m
```
Comment (jaycarlton) [inline .circleci/config.yml:469]: Can we define the URLs and ports as constants?

File: .circleci/config.yml [hunk 6]
```diff
@@ -264,136 +402,98 @@ jobs:
...
+    environment:
+      WORKBENCH_ENV: local
+    parallelism: 2
+    steps:
+      - attach_workspace:
+          at: .
+      - e2e_tests
+
+  # Run Puppeteer end-to-end UI tests on Workbench "test" environment.
+  e2e-test:
```
Comment (jaycarlton) [inline .circleci/config.yml:484]: The name should be `e2e-test-test` I think. Or make `e2e-test-local` just be `e2e-local`. It's unfortunate our qa environment is called test.

File: .circleci/config.yml [hunk 7]
```diff
@@ -264,136 +402,98 @@ jobs:
...
+    executor: browser_docker
+    environment:
+      WORKBENCH_ENV: dev
+    parallelism: 3
+    steps:
+      - run:
+          name: Sleep 1 minute before starting e2e tests
```
Comment (jaycarlton) [inline .circleci/config.yml:491]: I'd file a follow-on story to see if we can make this less absolute. Either we're going to wait longer than we need to, or timeout.

I bet there's either a user-submitted orb or a builtin feature for this situation.

File: .circleci/config.yml [hunk 8]
```diff
@@ -264,136 +402,98 @@ jobs:
...
-filter_releases: &filter_releases
-  filters:
-    branches:
-      ignore: /.*/
-    tags:
-      only: /^v.*/
 
 workflows:
-  version: 2
   build-test-deploy:
```
Comment (jaycarlton) [inline .circleci/config.yml:497]: Is there a way to give a workflow a human-readable name and description (in addition to the key `build-test-deploy`? Jobs and steps have them.

File: .circleci/config.yml [hunk 9]
```diff
@@ -22,76 +27,201 @@ anchors:
...
-          key: api-cache-{{ checksum "~/workbench/api/build.gradle" }}
 
-  api-local-test:
+  # See https://circleci.com/docs/2.0/workflows/#git-tag-job-execution
+  filter_master: &filter_master
+    filters:
+      branches:
+        only: master
+
+  filter_releases: &filter_releases
```
Comment (jaycarlton) [inline .circleci/config.yml:40]: Can you please add some comments here describing what these are for?

File: .circleci/config.yml [hunk 10]
```diff
@@ -22,76 +27,201 @@ anchors:
...
+          name: Update Git Submodules
+          command: git submodule update --init --recursive
+
+  ensure_branch_has_api_changes:
+    description: "Ensure branch has changes to API Code"
+    steps:
+      - run:
+          name: Ensure Branch has Changes to API Code
+          command: |
+            if [ ${CIRCLE_BRANCH} != "" ] && [ ${CIRCLE_BRANCH} != "master" ] && [ $(git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | grep api/ | wc -l | xargs) == 0 ]; then
```
Comment (jaycarlton) [inline .circleci/config.yml:114]: nit: if this gets any longer, can we move it to an include file or actually deploy it as a script? Shell code inside yaml is a bit difficult to read.

Comment (aweng98) [reply .circleci/config.yml:90]: original code. I don't want to change and break it.

File: .circleci/config.yml (hunk 1 above)
Comment (aweng98) [reply .circleci/config.yml:92]: cool.

File: .circleci/config.yml [hunk 11]
```diff
@@ -22,21 +31,217 @@ anchors:
...
+      - when:
+          condition: << parameters.restore >>
+          steps:
+            - restore_cache:
+                keys:
+                  - v1-api-cache-{{ checksum "~/workbench/api/build.gradle" }}-{{ .Branch }}
+                  - v1-api-cache-{{ checksum "~/workbench/api/build.gradle" }}-master
+                  - v1-api-cache-{{ checksum "~/workbench/api/build.gradle" }}
+
+  deploy_appengine:
```
Comment (aweng98) [inline .circleci/config.yml:174]: I'm not able to test any `deploy to app engine` jobs since they're only for master and release branches.

File: .circleci/config.yml [hunk 12]
```diff
@@ -71,176 +274,107 @@ jobs:
           - MYSQL_PASSWORD=ubuntu
     working_directory: ~/workbench
     environment:
-      # Drop to ~1.5G; the local API server starts two Java processes (the API
-      # server and the cron emulator). The default machine has 4GB, so we want
-      # the Java memory limit to be below half of that: other things on the
-      # machine need memory as well. Using 2G here manifested in 137 errors.
-      # If this limit proves to be too low, an alternative would be to use a
-      # larger resource class:
-      #   https://circleci.com/docs/2.0/configuration-reference/#resource_class
```
Comment (aweng98) [inline .circleci/config.yml:80]: with a `medium+` circleci container and `GRADLE_OPTS: "-Dorg.gradle.daemon=false"`, OOM issue should have been resolved. Thus, removed comments.

File: .circleci/config.yml (hunk 8 above)
Comment (aweng98) [reply .circleci/config.yml:466]: nope

File: .circleci/config.yml (hunk 10 above)
Comment (jaycarlton) [reply .circleci/config.yml]: Right, just wondering if you came across a way to do that.

File: .circleci/config.yml [hunk 13]
```diff
@@ -22,21 +31,198 @@ anchors:
...
+          command: yarn install --frozen-lockfile --non-interactive
+      - save_cache:
+          paths:
+            - ~/.cache/yarn
+            - e2e/node_modules
+          key: << parameters.version >>-yarn-e2e-cache-{{ checksum "~/workbench/e2e/yarn.lock" }}-{{ .Branch }}
+
+  api_cache:
+    description: "workbench/api: save and restore gradle cache"
+    parameters:
```
Comment (jaycarlton) [inline .circleci/config.yml:167]: So if they're both false, then nothing happens. It seems like one is required for meaningful operation. An accidental no-op command invocation could sit unnoticed for some time.

You could do a `cache_mode` parameter with an enum like `[SAVE, RESTORE, SAVE_AND_RESTORE]` and make that required? That would solve the potential no-op problem (though if you add a third command operation, it would get complicated).

Finally, you could break this down into very simple commands like `save_api_cache` and `restore_api_cache`, or even a `save_cache` with the name of the cache as a parameter.

File: .circleci/config.yml [hunk 14]
```diff
@@ -46,52 +232,41 @@ jobs:
           working_directory: ~/workbench/api
           command: ./project.rb gradle compile__bigquerytest__Java compile__integration__Java
       - run:
-          name: Run Java Unit Tests
+          name: Run Java unit tests
```
Comment (jaycarlton) [inline .circleci/config.yml:235]: +1 I like this sentence case much better.

File: .circleci/config.yml [hunk 15]
```diff
@@ -46,52 +232,41 @@ jobs:
...
   api-local-test:
+    # Refers to https://circleci.com/docs/2.0/configuration-reference/#resource_class
+    resource_class: medium+
     docker:
       - image: allofustest/workbench:buildimage-0.0.16
+      # Any known reason for not using circleci/mysql:5.7-ram? better performance and mitigates I/O contention.
+      # Refers to https://circleci.com/docs/2.0/circleci-images/#service-image-variant
       - image: mysql:5.7
         environment:
           - MYSQL_ROOT_PASSWORD=ubuntu
```
Comment (jaycarlton) [inline .circleci/config.yml:255]: Out of curiosity, have you tried the secrets cache? I realize these aren't secret, but might be an interesting thing to explore. We probably won't get permission to use it, though.

File: .circleci/config.yml [hunk 16]
```diff
@@ -105,142 +280,104 @@ jobs:
...
           path: ~/workbench/api/build/test-results/integrationTest/
-      - save_cache:
-          paths:
-            - ~/.gradle
-            - ~/.m2
-          key: api-integration-cache-{{ checksum "~/workbench/api/build.gradle" }}
+      - api_cache:
+          save: true
 
   api-nightly-integration-test:
```
Comment (jaycarlton) [inline .circleci/config.yml:336]: 👏 This is getting really tiny now!

File: .circleci/config.yml [hunk 17]
```diff
@@ -22,21 +31,198 @@ anchors:
...
+          command: |
+            if [ ${CIRCLE_BRANCH} != "" ] &&
+              [ ${CIRCLE_BRANCH} != "master" ] &&
+              [ $(git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | grep << parameters.dir_name >>/ | wc -l | xargs) == 0 ]; then
+                echo "No relevant changes in << parameters.dir_name >> directory in non-master branch. Skipping."
+                circleci step halt
+            fi
+          name: Ensure branch has changes to API or UI code
+
+  ui_dependencies:
```
Comment (jaycarlton) [inline .circleci/config.yml:116]: Please name commands like functions, i.e. starting with a verb. This one could be `install_ui_dependencies` I think. You could also start with `manage` but it's less descriptive.

File: .circleci/config.yml [hunk 18]
```diff
@@ -22,21 +31,198 @@ anchors:
...
+  activate_service_account_creds:
+    description: "Activiate CircleCI service account credential"
+    steps:
+      - run:
+          name: Activate CircleCI service account credential
+          working_directory: ~/workbench
+          # Used to call gsutil from the circle environment.
+          command: ci/activate_creds.sh api/circle-sa-key.json
+
+  git_checkout:
```
Comment (jaycarlton) [inline .circleci/config.yml:77]: nit: maybe name this `checkout_and_init_git`

File: .circleci/config.yml [hunk 19]
```diff
@@ -22,21 +31,198 @@ anchors:
...
+          name: "workbench/ui: yarn install dependencies"
+          working_directory: ~/workbench/ui
+          command: yarn install --verbose --frozen-lockfile --non-interactive
+      - save_cache:
+          paths:
+            - ~/.cache/yarn
+            - ui/node_modules
+          key: << parameters.version >>-yarn-ui-cache-{{ checksum "~/workbench/ui/yarn.lock" }}-{{ .Branch }}
+
+  e2e_dependencies:
```
Comment (jaycarlton) [inline .circleci/config.yml:143]: consider the name `install_e2e_dependencies`

File: .circleci/config.yml [hunk 20]
```diff
@@ -105,142 +280,104 @@ jobs:
...
       - deploy:
           name: Deploy to App Engine
           working_directory: ~/workbench/api
           command: |
             ./project.rb deploy \
               --project all-of-us-workbench-test \
-              --account circle-deploy-account@all-of-us-workbench-test.iam.gserviceaccount.com \
+              --account << pipeline.parameters.circleci_service_account >> \
               --version circle-ci-test \
               --key-file circle-sa-key.json \
```
Comment (jaycarlton) [inline .circleci/config.yml:306]: consider making this key file location a pipeline parameter as well.

File: .circleci/config.yml [hunk 21]
```diff
@@ -105,142 +280,104 @@ jobs:
...
               --promote
 
+
   api-deps-check:
     <<: *java_defaults
     steps:
-      - checkout
-      - run: *update_git_submodules
+      - git_checkout
       - run:
```
Comment (jaycarlton) [inline .circleci/config.yml:314]: Would it make sense to have a reusable scan command?

File: .circleci/config.yml [hunk 22]
```diff
@@ -105,142 +280,104 @@ jobs:
...
-          - api-integration-cache-
-      - run: *activate_service_accont_creds
+      - git_checkout
+      - api_cache:
+          restore: true
+      - activate_service_account_creds
       - run:
           name: Run Nightly Integration tests
           working_directory: ~/workbench/api
           command: ./project.rb nightly-integration
```
Comment (jaycarlton) [inline .circleci/config.yml:346]: So is the "nightly" test suite a different one than the regular one? Or is this purely a scheduling thing?

If it's just for scheduling, I bet we don't need separate jobs, but just different triggers and maybe parameters.

File: .circleci/config.yml [hunk 23]
```diff
@@ -22,21 +31,198 @@ anchors:
...
+          condition: << parameters.restore >>
+          steps:
+            - restore_cache:
+                keys:
+                  - v1-api-cache-{{ checksum "~/workbench/api/build.gradle" }}-{{ .Branch }}
+                  - v1-api-cache-{{ checksum "~/workbench/api/build.gradle" }}-master
+                  - v1-api-cache-{{ checksum "~/workbench/api/build.gradle" }}
+
+  run_e2e_test:
+    steps:
```
Comment (jaycarlton) [inline .circleci/config.yml:195]: Add a description here, please.

File: .circleci/config.yml [hunk 24]
```diff
@@ -22,21 +31,198 @@ anchors:
...
+      - run:
+          name: Export Puppeteer test environment variables
+          command: |
+            echo 'export USER_NAME=$PUPPETEER_TEST_USER' >> $BASH_ENV
+            echo 'export PASSWORD=$PUPPETEER_TEST_USER_PASSWORD' >> $BASH_ENV
+            echo 'export INVITATION_KEY=$PUPPETEER_TEST_REGISTRATION_KEY' >> $BASH_ENV
+            source $BASH_ENV
+      - run:
+          name: Run Puppeteer test
```
Comment (jaycarlton) [inline .circleci/config.yml:204]: nit: either call it e2e or Puppeteer throughout.

File: .circleci/config.yml [hunk 25]
```diff
@@ -46,52 +232,41 @@ jobs:
...
-          key: api-cache-{{ checksum "~/workbench/api/build.gradle" }}
+      - api_cache:
+          save: true
 
   api-local-test:
+    # Refers to https://circleci.com/docs/2.0/configuration-reference/#resource_class
+    resource_class: medium+
     docker:
       - image: allofustest/workbench:buildimage-0.0.16
+      # Any known reason for not using circleci/mysql:5.7-ram? better performance and mitigates I/O contention.
```
Comment (jaycarlton) [inline .circleci/config.yml:251]: Please file a Jira ticket for this comment to follow up. I don't think we need the comment here (it won't age well most likely).

File: .circleci/config.yml [hunk 26]
```diff
@@ -265,67 +402,38 @@ jobs:
           paths:
             - ui
 
-  # Run suite of Puppeteer end-to-end QA tests, pointed at the Workbench "test" environment.
+  # Run Puppeteer end-to-end UI tests on Workbench deployed "test" environment.
   puppeteer-e2e-test:
     <<: *defaults
```
Comment (jaycarlton) [inline .circleci/config.yml:407]: please add a job description

File: .circleci/config.yml [hunk 27]
```diff
@@ -265,67 +402,38 @@ jobs:
...
+      - run_e2e_test
+
   ui-deploy-to-test:
     <<: *defaults
     steps:
-      - checkout
-      - run: *update_git_submodules
-      # Use the ui-build-test workspace here to avoid redoing the setup.
+      - git_checkout
+      # Use "ui-build-test" job's workspace here to avoid redoing the setup.
```
Comment (jaycarlton) [inline .circleci/config.yml:425]: Can we not do this in an idempotent way with the cache facility? I'm nervous that we've introduced an order dependency here.

File: .circleci/config.yml [hunk 28]
```diff
@@ -337,63 +445,75 @@ jobs:
...
-filter_releases: &filter_releases
-  filters:
-    branches:
-      ignore: /.*/
-    tags:
-      only: /^v.*/
+
+  # Deploy local UI server and run Puppeteer 2e2 tests
+  e2e-local-ui:
+    <<: *defaults
```
Comment (jaycarlton) [inline .circleci/config.yml:488]: Please add a job description.

File: e2e/package.json [hunk 29]
```diff
@@ -9,7 +9,7 @@
     "test": "cross-env WORKBENCH_ENV=dev jest",
```
Comment (jaycarlton) [inline e2e/package.json:9]: Maybe comment somewhere on which values need to live in this file vs at the various levels in the CircleCI configs.

**Comment by jaycarlton** [comment]
> I am not able test out changes in deploy-staging, deploy-perf, api-deploy-to-test and ui-deploy-to-test because these jobs are filtered by filter_master.

Just go into a readme file and start rewording things or make yourself a SCRATCH.txt someplace and fiddle with that. 99% of the time that won't break anything.

File: .circleci/config.yml [hunk 30]
```diff
@@ -1,17 +1,26 @@
-version: 2
+version: 2.1
 
+# -------------------------
+#   PARAMETERS
+# -------------------------
+parameters:
+  cache_key_version:
```
Comment (calbach) [inline .circleci/config.yml:7]: I don't think this makes sense as a parameter, it is unlikely we're going to want to invalidate all of our caches simultaneously

File: .circleci/config.yml [hunk 31]
```diff
@@ -22,21 +31,203 @@ anchors:
...
+              [ ${CIRCLE_BRANCH} != "master" ] &&
+              ! git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | grep -qvE '(.md$)|(.pdf$)'; then
+                echo "No code changes in non-master branch. halting job."
+                circleci step halt
+            fi
+          name: Halt job and succeed early if no code changes on master branch
+
+  ensure_branch_has_changes:
+    description: "Ensure branch has changes to code in specified directory"
+    parameters:
```
Comment (calbach) [inline .circleci/config.yml:98]: Cool

File: .circleci/config.yml [hunk 32]
```diff
@@ -337,63 +448,74 @@ jobs:
...
+              --account << pipeline.parameters.circleci_service_account >> \
               --git-version "${CIRCLE_TAG}" \
               --app-version "${CIRCLE_TAG}" \
               --circle-url "https://circleci.com/workflow-run/${CIRCLE_WORKFLOW_ID}" \
               --key-file circle-sa-key.json \
               --promote
+          name: Deploy API and UI to "perf" App Engine
+
+  # Deploy local UI server and run Puppeteer 2e2 tests
+  puppeteer-e2e-local-ui:
```
Comment (calbach) [inline .circleci/config.yml:489]: This is definitely a new job - should this be a separate PR?

ALso, I thought the plan was to run a local UI and API server. Did you try to get this working?

File: .circleci/config.yml (hunk 12 above)
Comment (calbach) [reply .circleci/config.yml:80]: We need a comment above explaining why it has been set to medium+

File: .circleci/config.yml (hunk 32 above)
Comment (aweng98) [reply .circleci/config.yml]: Removed it from this PR. BTW, I couldn't get local ui/api working. I spent days on it, still having issues, I need your help!

File: .circleci/config.yml [hunk 33]
```diff
@@ -22,21 +31,202 @@ anchors:
...
+    branches:
+      only: master
+
+  # Job runs for no branch and only for tags starting with ‘v’
+  filter_only_release_tags: &filter_only_release_tags
+    branches:
+      ignore: /.*/
+    tags:
+      # Workbench release tag format example: v5-3-rc1
+      only: /^v[0-9]+-[0-9]+-rc[0-9]+$/
```
Comment (aweng98) [inline .circleci/config.yml:52]: regex short explanation:
`^` asserts position at start of a line.
`v` matches the character v literally (case sensitive).
`[0-9]+` match a single numerical digit (the `+` quantifier — matches between one and unlimited times, as many times as possible).
`-` matches the character `-` literally (case sensitive).
`rc` matches the character `rc` literally.
`$` asserts position at the end of a line.

Comment (jaycarlton) [reply .circleci/config.yml:52]: Awesome. Maybe put this in the comments? I doubt anyone will find this conversation again.

File: .circleci/config.yml [hunk 34]
```diff
@@ -22,21 +31,198 @@ anchors:
...
+          keys:
+            # restore cache strategy:
+            # Check for the latest cache from your branch.
+            # If that fails, then check for the latest from master branch.
+            # If that fails, then check for the latest from any branch
+            - << parameters.version >>-yarn-ui-cache-{{ checksum "~/workbench/ui/yarn.lock" }}-{{ .Branch }}
+            - << parameters.version >>-yarn-ui-cache-{{ checksum "~/workbench/ui/yarn.lock" }}-master
+            - << parameters.version >>-yarn-ui-cache-{{ checksum "~/workbench/ui/yarn.lock" }}
+      - run:
+          name: "workbench/ui: yarn install dependencies"
```
[the remark this answers is not in the document]
Comment (jaycarlton) [reply .circleci/config.yml:135]: The builtin ones are gerund title case, like `Making the Tea`.

File: .circleci/config.yml (hunk 32 above)
Comment (calbach) [reply .circleci/config.yml]: Sure thing, I didn't realize you were working on this. Please push up a draft PR of what you were trying and I can take a look.

File: .circleci/config.yml [hunk 35]
```diff
@@ -1,17 +1,26 @@
...
+#   PARAMETERS
+# -------------------------
+parameters:
+  cache_key_version:
+    type: string
+    default: "v1"
+
+  circleci_service_account:
+    type: string
+    default: "circle-deploy-account@all-of-us-workbench-test.iam.gserviceaccount.com"
```
[the remark this answers is not in the document]
Comment (aweng98) [reply .circleci/config.yml:9]: It can be hidden with either a [context](https://app.circleci.com/settings/organization/github/all-of-us/contexts?return-to=https%3A%2F%2Fapp.circleci.com%2Fpipelines%2Fgithub%2Fall-of-us%2Fworkbench%2F510%2Fworkflows%2Ff3fe3af4-2cb4-44bf-8640-803fd7765921%2Fjobs%2F79899%2Fsteps) or a [env variable](https://app.circleci.com/settings/project/github/all-of-us/workbench/environment-variables?return-to=https%3A%2F%2Fapp.circleci.com%2Fpipelines%2Fgithub%2Fall-of-us%2Fworkbench)
I can do change here or in another PR. Do you have any preference on how and when?

File: .circleci/config.yml [hunk 36]
```diff
@@ -22,21 +31,202 @@ anchors:
...
+              [ $(git diff --name-only $(git merge-base origin/master ${CIRCLE_BRANCH}) | grep << parameters.dir_name >>/ | wc -l | xargs) == 0 ]; then
+                echo "No relevant changes in << parameters.dir_name >> directory in non-master branch. Skipping."
+                circleci step halt
+            fi
+          name: Ensure branch has changes to code in specified directory
+
+  install_ui_dependencies:
+    description: "workbench/ui: yarn install, save and restore cache"
+    parameters:
+      version:
```
[the remark this answers is not in the document]
Comment (aweng98) [reply .circleci/config.yml:119]: no boolean eval in `<<   >>`, suggested changes involve shell scripts and `if`.
